File: packages/shared/domain/variables.ts
// P5 D20: the TypeScript mirrors of Go's model.Variable/model.Environment/model.VariableHistoryEntry
// and apivars.RevealResult. Following P2 D5's rule — "the wire shapes live in Go and are
// mirrored, not re-validated" — these are types, not guards; control.ts `trust<T>()`s them like
// every other bound result. Nothing here becomes tab state (unlike httpSavedRequestSchema in
// collections.ts), so none of it earns the one Zod-parse boundary that file's own comment explains.
//
// P12 D3: Http* → Api* — a variable is resolved for both protocols (Api* names the module).

export const VARIABLE_SCOPES = ['collection', 'environment'] as const;
export type VariableScope = (typeof VARIABLE_SCOPES)[number];

// D12: the row model, deliberately without `enabled` or `type` — see the plan's own table for why
// each is left out. `value` is '' whenever `isSecret` — a secret's plaintext never crosses the
// bridge except through the gated Reveal (D5). `description` is P17 D14 — a plain field, not a
// secret, added alongside the four already here.
export interface ApiVariable {
  id: string;
  scope: VariableScope;
  ownerId: string;
  name: string;
  value: string;
  isSecret: boolean;
  sortOrder: number;
  description: string;
}

// D3: the active environment is app-global, a column on the row itself rather than a layout leaf
// or a settings key. `description` is P17 D14 — app-local free text, no Postman round-trip
// question at all (unlike a variable's, F10).
export interface ApiEnvironment {
  id: string;
  name: string;
  sortOrder: number;
  isActive: boolean;
  description: string;
}

// D13: history is per-entry — an environment has no history of its own, only its entries do.
export interface ApiVariableHistoryEntry {
  id: string;
  variableId: string;
  value: string;
  isSecret: boolean;
  recordedAt: string;
}

// P17 D21/D22/D23: one line of a parsed `.env` bulk edit — the wire mirror of
// api-core's own dotenv.ts#EnvEntry, carried across the bridge to VariablesRepo.ApplyBulk.
// hasValue is false for a bare `KEY=` line (D22 rule 3).
export interface ApiVariableBulkEntry {
  name: string;
  value: string;
  hasValue: boolean;
  description: string;
}

// ApplyBulk's own answer — the same counts the renderer's own reconcileEnv already computed
// locally for its live summary (§4's guard that the two reconciles agree).
export interface ApiVariableBulkResult {
  added: number;
  updated: number;
  removed: number;
  reordered: boolean;
}

// D8: the same four-outcome vocabulary v1.1 P14 already established for connections.RevealResult,
// redeclared here rather than shared — importing internal/connections from internal/apivars would
// be exactly the Studio<->Api coupling this app's module-boundary rule exists to prevent.
export type RevealOutcome = 'revealed' | 'cancelled' | 'confirmation-required' | 'error';

export interface RevealResult {
  value: string | null;
  error: string | null;
  outcome: RevealOutcome;
}

File: apps/kira-studio/internal/storage/model/variables.go
package model

import "fmt"

// VariableScope discriminates which owner an api_variables row belongs to. It is computed on
// scan from which foreign key is non-null (P5 D4) — never a stored column, so there is no second
// source of the same fact to drift from the CHECK the migration already enforces.
type VariableScope string

const (
	VariableScopeCollection  VariableScope = "collection"
	VariableScopeEnvironment VariableScope = "environment"
)

// Environment is one api_environments row (P5 D3/D4). IsActive is the app-global selection — the
// repo guarantees at most one row carries it. Description is P17 D14 — app-local free text, no
// Postman round-trip question at all (unlike a variable's, F10). Color is P18 D16/D19 — a
// user-assigned value from the same palette Studio's connections use (ValidPaletteColor), never
// derived; 'none' is the default.
type Environment struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	SortOrder   int    `json:"sortOrder"`
	IsActive    bool   `json:"isActive"`
	Description string `json:"description"`
	Color       string `json:"color"`
}

// Variable is one api_variables row's **list projection** — P5 D4/D5. Value is ” whenever
// IsSecret, and the row's own encrypted-secret column is not a field of this struct at all, so
// "the list never returns a secret's plaintext or ciphertext" is a property of this type, not of a
// per-row branch somewhere that could later be forgotten. Description is P17 D14 — not a secret,
// so it does not weaken that property.
type Variable struct {
	ID          string        `json:"id"`
	Scope       VariableScope `json:"scope"`
	OwnerID     string        `json:"ownerId"`
	Name        string        `json:"name"`
	Value       string        `json:"value"`
	IsSecret    bool          `json:"isSecret"`
	SortOrder   int           `json:"sortOrder"`
	Description string        `json:"description"`
}

// VariableHistoryEntry is one api_variable_history row's list projection — the same "no secret
// column here" discipline as Variable above. apivars.Service.RevealHistory is the only path to a
// secret history entry's plaintext.
type VariableHistoryEntry struct {
	ID         string `json:"id"`
	VariableID string `json:"variableId"`
	Value      string `json:"value"`
	IsSecret   bool   `json:"isSecret"`
	RecordedAt string `json:"recordedAt"`
}

// Validate checks what SQL cannot: a non-empty name. D12: a duplicate name within one scope is
// allowed (resolved first-wins by sort_order in the renderer) — this is the only invariant the
// repo does not otherwise enforce through a constraint.
func (v Variable) Validate() error {
	if v.Name == "" {
		return fmt.Errorf("model: variable: name is required")
	}
	return nil
}

// VariableBulkEntry is one line of a parsed `.env` bulk edit (P17 D21/D22/D23) — the renderer's
// own dotenv.ts `EnvEntry`, carried across the bridge. HasValue is false for a bare `KEY=` line
// (D22 rule 3: a secret row whose line carries no value keeps its stored secret untouched).
type VariableBulkEntry struct {
	Name        string `json:"name"`
	Value       string `json:"value"`
	HasValue    bool   `json:"hasValue"`
	Description string `json:"description"`
}

// VariableBulkResult is ApplyBulk's own answer — the same counts the renderer's own
// dotenv.ts#reconcileEnv already computed locally for its live summary, so returning them here is
// also the guard that the two reconciles agree (§4 of the plan).
type VariableBulkResult struct {
	Added     int  `json:"added"`
	Updated   int  `json:"updated"`
	Removed   int  `json:"removed"`
	Reordered bool `json:"reordered"`
}
